File: src/parser/parser.c
#ifndef PARSER_H
#define PARSER_H

#include "../std/Allocator.c"
#include "../std/Vec.c"
#include "tokenizer.c"

typedef enum {
  CallExpression,
  IdentifierExpression,
  NumericExpression,
  StringExpression,
  ArithmeticExpression,
} ExpressionType;

typedef struct Expression {
  ExpressionType type;
  union {
    struct {
      struct Expression *callee;
      struct Expression *parameters;
      size_t parameters_len;
    } call;
    Slice(char) number;
    Slice(char) string;
    Slice(char) identifier;
    struct {
      char op;
      struct Expression *left;
      struct Expression *right;
    } arithmetic;
  };
} Expression;

DefSlice(Expression);
DefResult(Slice_Expression);
DefVec(Expression);
DefResult(Vec_Expression);

typedef struct {
  Slice(char) name;
  Expression value;
  bool constant;
} Declaration;

typedef struct {
  Slice(char) name;
  Expression value;
} Assignment;

typedef enum {
  StatementEOF = 0,
  ExpressionStatement,
  DeclarationStatement,
  AssignmentStatement,
  InlineBatchStatement,
  BlockStatement,
  IfStatement,
} StatementType;

typedef struct If If;
typedef struct Block Block;

typedef struct {
  StatementType type;
  union {
    Expression expression;
    Declaration declaration;
    Assignment assignment;
    Slice(char) inline_batch;
    If *if_statement;
    Block *block;
  };
} Statement;

DefSlice(Statement);
DefResult(Slice_Statement);
DefVec(Statement);
DefResult(Vec_Statement);

struct If {
  Expression condition;
  Statement *consequence;
  Statement *alternate;
};

DefSlice(If);
DefResult(Slice_If);

struct Block {
  Slice(Statement) statements;
};

DefSlice(Block);
DefResult(Slice_Block);

static void printExpression(Expression expr) {
  fprintf(stdout, "Expr:");
  switch (expr.type) {
  case CallExpression: {
    fprintf(stdout, "Call:(");
    printExpression(*expr.call.callee);
    fprintf(stdout, ") with (");
    if (expr.call.parameters_len > 0) {
      printExpression(expr.call.parameters[0]);
    }
    for (size_t i = 1; i < expr.call.parameters_len; i++) {
      fprintf(stdout, ", ");
      printExpression(expr.call.parameters[i]);
    }
    fprintf(stdout, ")");
  } break;
  case IdentifierExpression: {
    fprintf(stdout, "Ident(%1.*s)", (int)expr.identifier.len,
            expr.identifier.ptr);
  } break;
  case NumericExpression: {
    fprintf(stdout, "Number(%1.*s)", (int)expr.number.len, expr.number.ptr);
  } break;
  case StringExpression: {
    fprintf(stdout, "String(\"%1.*s\")", (int)expr.string.len, expr.string.ptr);
  } break;
  case ArithmeticExpression: {

    fprintf(stdout, "Arith(");
    printExpression(*expr.arithmetic.left);
    fprintf(stdout, " %c ", expr.arithmetic.op);
    printExpression(*expr.arithmetic.right);
    fprintf(stdout, ")");
  } break;
  }
}

static void printStatement(Statement stmt) {
  switch (stmt.type) {
  case ExpressionStatement: {
    Expression expr = stmt.expression;
    printExpression(expr);
    fprintf(stdout, "\n");
  } break;
  case DeclarationStatement: {
    Declaration decl = stmt.declaration;
    fprintf(stdout, "%1.*s :%c ", (int)decl.name.len, decl.name.ptr,
            decl.constant ? ':' : '=');
    printExpression(decl.value);
    fprintf(stdout, "\n");
  } break;
  case AssignmentStatement: {
    Assignment assign = stmt.assignment;
    fprintf(stdout, "%1.*s = ", (int)assign.name.len, assign.name.ptr);
    printExpression(assign.value);
    fprintf(stdout, "\n");
  } break;
  case InlineBatchStatement: {
    fprintf(stdout, "Inline Batch {\n");
    fprintf(stdout, "%1.*s", (int)stmt.inline_batch.len, stmt.inline_batch.ptr);
    fprintf(stdout, "}\n");
  } break;
  case IfStatement: {
    fprintf(stdout, "If (");
    printExpression(stmt.if_statement->condition);
    fprintf(stdout, ") ");
    printStatement(*stmt.if_statement->consequence);
    if (stmt.if_statement->alternate) {
      fprintf(stdout, " else ");
      printStatement(*stmt.if_statement->alternate);
    }
  } break;
  case BlockStatement: {
    fprintf(stdout, "Block {\n");
    for (size_t i = 0; i < stmt.block->statements.len; i++) {
      printStatement(stmt.block->statements.ptr[i]);
    }
    fprintf(stdout, "}\n");
  } break;
  case StatementEOF: {
    panic("StatementEOF");
  } break;
  }
}

typedef struct {
  Slice(Statement) statements;
} Program;

// static Expression parseUnitExpression(Allocator ally, TokenIterator *it,
//                                       Token t) {}

static inline Expression parseExpression(Allocator ally, TokenIterator *it,
                                         Token t) {
  switch (t.type) {
  case TokenType_Number: {
    Token next = peekToken(it);
    if (next.type != TokenType_Star && next.type != TokenType_Plus &&
        next.type != TokenType_Hyphen && next.type != TokenType_Slash &&
        next.type != TokenType_Equal) {
      return (Expression){.type = NumericExpression, .number = t.number};
    }
    if (nextToken(it).type == TokenType_Equal) { // op
      if (peekToken(it).type != TokenType_Equal) {
        panic("Invalid expression following <num> =");
      }
      // ==
      //  ^
      nextToken(it);
    }

    Result(Slice_Expression) lr_res = alloc(ally, Expression, 2);
    if (!lr_res.ok)
      panic(lr_res.err);
    Expression *left = &lr_res.val.ptr[0];
    Expression *right = &lr_res.val.ptr[1];
    *left = parseExpression(ally, it, t);
    *right = parseExpression(ally, it, nextToken(it));
    return (Expression){
        .type = ArithmeticExpression,
        .arithmetic =
            {
                .op = next.type == TokenType_Star     ? '*'
                      : next.type == TokenType_Plus   ? '+'
                      : next.type == TokenType_Hyphen ? '-'
                      : next.type == TokenType_Equal  ? '=' // comparison
                                                      : '/',
                .left = left,
                .right = right,
            },
    };
  }
  case TokenType_String:
    return (Expression){.type = StringExpression, .string = t.string};
  case TokenType_Ident: {
    Token next = peekToken(it);
    if (next.type == TokenType_Star || next.type == TokenType_Plus ||
        next.type == TokenType_Hyphen || next.type == TokenType_Slash ||
        next.type == TokenType_Equal) {
      if (nextToken(it).type == TokenType_Equal) { // op
        if (peekToken(it).type != TokenType_Equal) {
          panic("Invalid expression following <num> =");
        }
        // ==
        //  ^
        nextToken(it);
      }
      Result(Slice_Expression) lr_res = alloc(ally, Expression, 2);
      if (!lr_res.ok)
        panic(lr_res.err);
      Expression *left = &lr_res.val.ptr[0];
      Expression *right = &lr_res.val.ptr[1];
      *left = (Expression){
          .type = IdentifierExpression,
          .number = t.ident,
      };
      *right = parseExpression(ally, it, nextToken(it));
      return (Expression){
          .type = ArithmeticExpression,
          .arithmetic =
              {
                  .op = next.type == TokenType_Star     ? '*'
                        : next.type == TokenType_Plus   ? '+'
                        : next.type == TokenType_Hyphen ? '-'
                        : next.type == TokenType_Equal  ? '=' // comparison
                                                        : '/',
                  .left = left,
                  .right = right,
              },
      };
    }
    if (next.type == TokenType_OpenParen) {
      // call expression
      nextToken(it);
      Result(Vec_Expression) res = createVec(ally, Expression, 1);
      if (!res.ok) {
        panic(res.err);
      }
      Vec(Expression) parameters = res.val;
      Token param = nextToken(it);
      while (param.type != TokenType_CloseParen) {
        if (param.type == TokenType_EOF) {
          panic("parseExpression: Unclosed open paren");
        }
        Expression expr = parseExpression(ally, it, param);
        Token paramSep = nextToken(it);
        if (paramSep.type != TokenType_Comma &&
            paramSep.type != TokenType_CloseParen) {
          panic("parseExpression: Parameter list expression not followed by "
                "comma or close paren ^");
        }
        if (!append(&parameters, Expression, &expr)) {
          panic("Failed to append to parameter list");
        }
        if (paramSep.type == TokenType_CloseParen) {
          break;
        }
        param = nextToken(it);
      }
      shrinkToLength(&parameters, Expression);

      Result(Slice_Expression) callee_res = alloc(ally, Expression, 1);
      if (!callee_res.ok) {
        panic("parseExpression: Failed to allocate callee");
      }
      Expression *callee = callee_res.val.ptr;
      *callee = (Expression){
          .type = IdentifierExpression,
          .identifier = t.ident,
      };

      return (Expression){.type = CallExpression,
                          .call = {.callee = callee,
                                   .parameters = parameters.slice.ptr,
                                   .parameters_len = parameters.slice.len}};
    }
    // identifier expression
    return (Expression){.type = IdentifierExpression, .identifier = t.ident};
  }
  case TokenType_EOF:
  case TokenType_OpenParen:
  case TokenType_CloseParen:
  case TokenType_OpenCurly:
  case TokenType_CloseCurly:
  case TokenType_Semi:
  case TokenType_Comma:
  case TokenType_Colon:
  case TokenType_Equal:
  case TokenType_Star:
  case TokenType_Plus:
  case TokenType_Hyphen:
  case TokenType_Slash:
  case TokenType_InlineBatch:
  case TokenType_Unknown: {
    printToken(t);
    panic("\nparseExpression: Invalid TokenType ^");
  }
  }
}

static Statement parseStatement(Allocator ally, TokenIterator *it) {
  TokenIterator snapshot = *it;
  Token t = nextToken(it);

  switch (t.type) {

  case TokenType_Ident:
  case TokenType_Number:
  case TokenType_String: {
    if (t.type == TokenType_Ident && peekToken(it).type == TokenType_Colon) {
      nextToken(it); // :
      Token afterColon = peekToken(it);
      if (afterColon.type != TokenType_Equal &&
          afterColon.type != TokenType_Colon) {
        printToken(peekToken(it));
        panic("Invalid token following colon ^");
      }
      nextToken(it); // =
      Expression value = parseExpression(ally, it, nextToken(it));

      Statement decl_stmt = {
          .type = DeclarationStatement,
          .declaration = {.name = t.ident,
                          .value = value,
                          .constant = afterColon.type == TokenType_Colon},
      };
      Token semi = nextToken(it);
      if (semi.type != TokenType_Semi) {
        printToken(semi);
        panic("\nparse: Unknown token following expression statement ^");
      }
      return decl_stmt;
    } else if (t.type == TokenType_Ident &&
               eql(t.ident, (Slice(char)){.ptr = "if", .len = 2})) {
      if (peekToken(it).type != TokenType_OpenParen) {
        panic("Missing ( after if");
      }
      nextToken(it); // (
      Expression condition = parseExpression(ally, it, nextToken(it));
      if (peekToken(it).type != TokenType_CloseParen) {
        printToken(peekToken(it));
        panic("\nMissing ) after if condition");
      }
      nextToken(it); // )
      Result(Slice_If) if_res = alloc(ally, If, 1);
      if (!if_res.ok)
        panic(if_res.err);
      If *if_statement = if_res.val.ptr;
      Result(Slice_Statement) cons_res = alloc(ally, Statement, 1);
      if (!cons_res.ok)
        panic(cons_res.err);
      Statement *consequence = cons_res.val.ptr;
      *consequence = parseStatement(ally, it);
      if_statement->condition = condition;
      if_statement->consequence = consequence;
      if_statement->alternate = NULL;
      Token elseToken = peekToken(it);
      if (elseToken.type == TokenType_Ident &&
          eql(elseToken.ident, (Slice(char)){.ptr = "else", .len = 4})) {
        nextToken(it);
        Result(Slice_Statement) alt_res = alloc(ally, Statement, 1);
        if (!alt_res.ok)
          panic(alt_res.err);
        Statement *alternate = alt_res.val.ptr;
        *alternate = parseStatement(ally, it);
        if_statement->alternate = alternate;
      }
      Statement s = {
          .type = IfStatement,
          .if_statement = if_statement,
      };
      return s;
    } else if (t.type == TokenType_Ident &&
               peekToken(it).type == TokenType_Equal) {
      nextToken(it);
      Expression value = parseExpression(ally, it, nextToken(it));

      Statement assign_stmt = {
          .type = AssignmentStatement,
          .assignment = {.name = t.ident, .value = value},
      };
      Token semi = nextToken(it);
      if (semi.type != TokenType_Semi) {
        printToken(semi);
        panic("\nparse: Unknown token following expression statement ^");
      }
      return assign_stmt;
    } else {
      Statement s = {
          .type = ExpressionStatement,
          .expression = parseExpression(ally, it, t),
      };
      Token semi = nextToken(it);
      if (semi.type != TokenType_Semi) {
        printToken(semi);
        panic("\nparse: Unknown token following expression statement ^");
      }
      return s;
    }
  } break;
  case TokenType_InlineBatch: {
    Statement s = {
        .type = InlineBatchStatement,
        .inline_batch = t.inline_batch,
    };
    return s;
  } break;
  case TokenType_OpenCurly: {
    Result(Vec_Statement) statements_res = createVec(ally, Statement, 4);
    if (!statements_res.ok)
      panic(statements_res.err);
    Vec(Statement) statements = statements_res.val;
    Statement stmt = parseStatement(ally, it);
    while (stmt.type) {
      printStatement(stmt);
      if (!append(&statements, Statement, &stmt)) {
        panic("Failed to append statement in block");
      }
      stmt = parseStatement(ally, it);
    }
    Token closecurly = peekToken(it);
    if (closecurly.type != TokenType_CloseCurly) {
      printToken(closecurly);
      printf("fart\n");
      panic("\nparse: Unknown token following block ^");
    }
    nextToken(it); // }
    Result(Slice_Block) block_res = alloc(ally, Block, 1);
    if (!block_res.ok)
      panic(block_res.err);
    block_res.val.ptr->statements = statements.slice;
    return (Statement){.type = BlockStatement, .block = block_res.val.ptr};
  } break;
  case TokenType_EOF:
  case TokenType_OpenParen:
  case TokenType_CloseParen:
  case TokenType_CloseCurly:
  case TokenType_Semi:
  case TokenType_Comma:
  case TokenType_Colon:
  case TokenType_Equal:
  case TokenType_Star:
  case TokenType_Plus:
  case TokenType_Hyphen:
  case TokenType_Slash:
  case TokenType_Unknown: {
    *it = snapshot; // restore
    return (Statement){.type = StatementEOF};
  } break;
  }
}

static Program parse(Allocator ally, TokenIterator *it) {
  Result(Vec_Statement) res = createVec(ally, Statement, 16);
  if (!res.ok) {
    panic("parse: Failed to alloc statements");
  }
  Vec(Statement) statements = res.val;

  Statement stmt = parseStatement(ally, it);
  while (stmt.type) {
    if (!append(&statements, Statement, &stmt)) {
      panic("Failed to append to statement list");
    }
    stmt = parseStatement(ally, it);
  }

  shrinkToLength(&statements, Statement);
  return (Program){.statements = statements.slice};
}

#endif /* PARSER_H */

File: src/parser/parser.zig
const std = @import("std");
const builtin = @import("builtin");
const Lexer = @import("Lexer.zig");

const ExpressionType = enum(c_int) {
    call,
    identifier,
    numeric,
    string,
    arithmetic,
    function,
};

pub const Expression = struct {
    tag: ExpressionType,
    x: union {
        call: struct {
            callee: *Expression,
            parameters: [*]const Expression,
            parameters_len: usize,
        },
        number: []const u8,
        string: []const u8,
        identifier: []const u8,
        arithmetic: struct {
            op: u8,
            left: *Expression,
            right: *Expression,
        },
        function_expression: struct {
            parameters: [*]const Expression,
            parameters_len: usize,
            body: *Statement,
        },
    },
};

const Declaration = struct {
    name: []const u8,
    value: Expression,
    constant: bool,
};

const Assignment = struct {
    name: []const u8,
    value: Expression,
};

pub const StatementType = enum(c_int) {
    eof,
    expression,
    declaration,
    assignment,
    inline_batch,
    block,
    @"if",
    @"while",
    @"return",
};

pub const Statement = struct {
    tag: StatementType,
    x: union {
        expression: Expression,
        declaration: Declaration,
        assignment: Assignment,
        inline_batch: []const u8,
        @"if": *If,
        @"while": *While,
        block: *Block,
        @"return": ?*Expression,
    } = undefined,
};

pub const If = struct {
    condition: Expression,
    consequence: *Statement,
    alternate: ?*Statement,
};

const While = struct {
    condition: Expression,
    body: *Statement,
};

pub const Block = struct {
    statements: []const Statement,
};

pub fn printExpression(expr: Expression) (error{StatementEOF} || std.fs.File.WriteError)!void {
    const stdout = std.io.getStdOut().writer();
    try stdout.writeAll("Expr:");
    switch (expr.tag) {
        .call => {
            try stdout.writeAll("Call:(");
            try printExpression(expr.x.call.callee.*);
            try stdout.writeAll(") with (");
            if (expr.x.call.parameters_len > 0) {
                try printExpression(expr.x.call.parameters[0]);
            }
            for (1..expr.x.call.parameters_len) |i| {
                try stdout.writeAll(", ");
                try printExpression(expr.x.call.parameters[i]);
            }
            try stdout.writeAll(")");
        },
        .identifier => {
            try stdout.print("Ident({s})", .{expr.x.identifier});
        },
        .numeric => {
            try stdout.print("Number({s})", .{expr.x.number});
        },
        .string => {
            try stdout.print("String(\"{s}\")", .{expr.x.string});
        },
        .arithmetic => {
            try stdout.writeAll("Arith(");
            try printExpression(expr.x.arithmetic.left.*);
            try stdout.writeByte(expr.x.arithmetic.op);
            try printExpression(expr.x.arithmetic.right.*);
            try stdout.writeByte(')');
        },
        .function => {
            try stdout.writeAll("Function (");
            if (expr.x.function_expression.parameters_len > 0) {
                try printExpression(expr.x.function_expression.parameters[0]);
            }
            for (1..expr.x.function_expression.parameters_len) |i| {
                try stdout.writeAll(", ");
                try printExpression(expr.x.function_expression.parameters[i]);
            }
            try stdout.writeAll(") ");
            try printStatement(expr.x.function_expression.body.*);
        },
    }
}

pub fn printStatement(stmt: Statement) (error{StatementEOF} || std.fs.File.WriteError)!void {
    const stdout = std.io.getStdOut().writer();
    switch (stmt.tag) {
        .expression => {
            try printExpression(stmt.x.expression);
            try stdout.writeByte('\n');
        },
        .declaration => {
            const decl = stmt.x.declaration;
            try stdout.print("{s} :{c} ", .{
                decl.name,
                @as(u8, if (decl.constant) ':' else '='),
            });
            try printExpression(decl.value);
            try stdout.writeByte('\n');
        },
        .assignment => {
            const assign = stmt.x.assignment;
            try stdout.print("{s} = ", .{assign.name});
            try printExpression(assign.value);
            try stdout.writeByte('\n');
        },
        .inline_batch => {
            try stdout.writeAll("Inline Batch {\n");
            try stdout.writeAll(stmt.x.inline_batch);
            try stdout.writeAll("}\n");
        },
        .@"if" => {
            try stdout.writeAll("If (");
            try printExpression(stmt.x.@"if".condition);
            try stdout.writeAll(") ");
            try printStatement(stmt.x.@"if".consequence.*);
            if (stmt.x.@"if".alternate) |alt| {
                try stdout.writeAll(" else ");
                try printStatement(alt.*);
            }
        },
        .@"while" => {
            try stdout.writeAll("While (");
            try printExpression(stmt.x.@"while".condition);
            try stdout.writeAll(") ");
            try printStatement(stmt.x.@"while".body.*);
        },
        .block => {
            try stdout.writeAll("Block {\n");
            for (stmt.x.block.statements.ptr[0..stmt.x.block.statements.len]) |s| {
                try printStatement(s);
            }
            try stdout.writeAll("}\n");
        },
        .@"return" => {
            try stdout.writeAll("Return (");
            if (stmt.x.@"return") |expr| try printExpression(expr.*);
            try stdout.writeAll(")\n");
        },
        .eof => return error.StatementEOF,
    }
}

pub const Program = struct {
    statements: []const Statement,
};

pub fn parseParameters(ally: std.mem.Allocator, it: *Lexer) ![]const Expression {
    var parameters = try std.ArrayList(Expression).initCapacity(ally, 1);
    var param = it.next().?;
    while (param != .closeParen) : (param = it.next().?) {
        const expr = try parseExpression(ally, it, param);
        const paramSep = it.next().?;
        if (paramSep != .comma and paramSep != .closeParen) {
            @panic("parseExpression: Parameter list expression not followed by comma or close paren ^");
        }
        try parameters.append(expr);
        if (paramSep == .closeParen) break;
    }
    return parameters.toOwnedSlice();
}

pub fn parseExpression(ally: std.mem.Allocator, it: *Lexer, t: Lexer.Token) (error{StatementEOF} || std.mem.Allocator.Error || std.fs.File.WriteError)!Expression {
    switch (t) {
        .number => |number| {
            const next = Lexer.peek(it).?;
            if (next != .star and
                next != .plus and
                next != .excl and
                next != .hyphen and
                next != .slash and
                next != .percent and
                next != .equal)
                return .{ .tag = .numeric, .x = .{ .number = number } };
            _ = it.next().?;
            if (next == .equal or next == .excl) {
                if (Lexer.peek(it).? != .equal) {
                    @panic("Invalid expression following <num> =");
                }
                // ==
                //  ^
                _ = it.next().?;
            }
            const lr = try ally.alloc(Expression, 2);
            const left = &lr[0];
            const right = &lr[1];
            left.* = try parseExpression(ally, it, t);
            right.* = try parseExpression(ally, it, it.next().?);
            return .{
                .tag = .arithmetic,
                .x = .{ .arithmetic = .{
                    .op = switch (next) {
                        .star => '*',
                        .plus => '+',
                        .hyphen => '-',
                        .equal => '=',
                        .excl => '!',
                        .percent => '%',
                        else => '/',
                    },
                    .left = left,
                    .right = right,
                } },
            };
        },
        .string => |string| {
            return .{ .tag = .string, .x = .{ .string = string } };
        },
        .ident => |ident| {
            const next = Lexer.peek(it).?;
            if (next == .star or
                next == .plus or
                next == .excl or
                next == .hyphen or
                next == .slash or
                next == .percent or
                next == .equal)
            {
                _ = it.next().?;
                if (next == .equal or next == .excl) {
                    if (Lexer.peek(it).? != .equal) {
                        @panic("Invalid expression following <num> =");
                    }
                    // ==
                    //  ^
                    _ = it.next().?;
                }
                const lr = try ally.alloc(Expression, 2);
                const left = &lr[0];
                const right = &lr[1];
                left.* = .{
                    .tag = .identifier,
                    .x = .{ .identifier = ident },
                };
                right.* = try parseExpression(ally, it, it.next().?);
                return .{
                    .tag = .arithmetic,
                    .x = .{ .arithmetic = .{
                        .op = switch (next) {
                            .star => '*',
                            .plus => '+',
                            .hyphen => '-',
                            .equal => '=',
                            .excl => '!',
                            .percent => '%',
                            else => '/',
                        },
                        .left = left,
                        .right = right,
                    } },
                };
            }
            if (next == .openParen) {
                // call expression
                _ = it.next().?;
                const parameters = try parseParameters(ally, it);
                const callee = try ally.create(Expression);
                callee.* = .{
                    .tag = .identifier,
                    .x = .{ .identifier = ident },
                };
                return .{
                    .tag = .call,
                    .x = .{ .call = .{
                        .callee = callee,
                        .parameters = parameters.ptr,
                        .parameters_len = parameters.len,
                    } },
                };
            }
            return .{ .tag = .identifier, .x = .{ .identifier = ident } };
        },
        .openParen => {
            const parameters = try parseParameters(ally, it);
            const body = try ally.create(Statement);
            body.* = try parseStatement(ally, it);
            return .{
                .tag = .function,
                .x = .{ .function_expression = .{
                    .parameters = parameters.ptr,
                    .parameters_len = parameters.len,
                    .body = body,
                } },
            };
        },
        .closeParen,
        .openCurly,
        .closeCurly,
        .semi,
        .comma,
        .colon,
        .equal,
        .excl,
        .star,
        .plus,
        .hyphen,
        .slash,
        .percent,
        .inline_batch,
        .unknown,
        => {
            try std.io.getStdOut().writer().print("{}", .{t});
            @panic("\nparseExpression: Invalid TokenType ^");
        },
    }
}

pub fn parseStatement(ally: std.mem.Allocator, it: *Lexer) !Statement {
    const snapshot = it.*;
    const t = it.next() orelse {
        it.* = snapshot; // restore
        return .{ .tag = .eof };
    };

    switch (t) {
        .ident, .number, .string => |str| {
            if (t == .ident and std.mem.eql(u8, str, "if")) {
                if (Lexer.peek(it).? != .openParen) {
                    @panic("Missing ( after if");
                }
                _ = it.next().?; // (
                const condition = try parseExpression(ally, it, it.next().?);
                if (Lexer.peek(it).? != .closeParen) {
                    try Lexer.peek(it).?.print();
                    @panic("\nMissing ) after if condition");
                }
                _ = it.next().?; // )
                const if_statement = try ally.create(If);
                const consequence = try ally.create(Statement);
                consequence.* = try parseStatement(ally, it);
                if_statement.condition = condition;
                if_statement.consequence = consequence;
                if_statement.alternate = null;
                const elseToken = Lexer.peek(it).?;
                if (elseToken == .ident and std.mem.eql(u8, elseToken.ident, "else")) {
                    _ = it.next().?;
                    const alternate = try ally.create(Statement);
                    alternate.* = try parseStatement(ally, it);
                    if_statement.alternate = alternate;
                }
                return .{ .tag = .@"if", .x = .{ .@"if" = if_statement } };
            } else if (t == .ident and std.mem.eql(u8, str, "while")) {
                if (Lexer.peek(it).? != .openParen) {
                    @panic("Missing ( after while");
                }
                _ = it.next().?; // (
                const condition = try parseExpression(ally, it, it.next().?);
                if (Lexer.peek(it).? != .closeParen) {
                    try Lexer.peek(it).?.print();
                    @panic("\nMissing ) after while condition");
                }
                _ = it.next().?; // )
                const while_statement = try ally.create(While);
                const body = try ally.create(Statement);
                body.* = try parseStatement(ally, it);
                while_statement.condition = condition;
                while_statement.body = body;
                return .{ .tag = .@"while", .x = .{ .@"while" = while_statement } };
            } else if (t == .ident and std.mem.eql(u8, str, "return")) {
                if (Lexer.peek(it).? == .semi) {
                    _ = it.next().?;
                    return .{ .tag = .@"return", .x = .{ .@"return" = null } };
                }
                const ret = try ally.create(Expression);
                ret.* = try parseExpression(ally, it, it.next().?);
                const semi = it.next().?;
                if (semi != .semi) {
                    try semi.print();
                    @panic("\nparse: Unknown Lexeren following expression statement ^");
                }
                return .{ .tag = .@"return", .x = .{ .@"return" = ret } };
            } else if (t == .ident and Lexer.peek(it).? == .colon) {
                _ = it.next().?; // :
                const afterColon = Lexer.peek(it).?;
                if (afterColon != .equal and afterColon != .colon) {
                    try Lexer.peek(it).?.print();
                    @panic("Invalid Lexeren following colon ^");
                }
                _ = it.next().?; // =
                const value = try parseExpression(ally, it, it.next().?);
                const decl_stmt = Statement{
                    .tag = .declaration,
                    .x = .{ .declaration = .{
                        .name = str,
                        .value = value,
                        .constant = afterColon == .colon,
                    } },
                };
                const semi = it.next().?;
                if (semi != .semi) {
                    try semi.print();
                    @panic("\nparse: Unknown Lexeren following expression statement ^");
                }
                return decl_stmt;
            } else if (t == .ident and Lexer.peek(it).? == .equal) {
                _ = it.next().?;
                const value = try parseExpression(ally, it, it.next().?);
                const assign_stmt = Statement{
                    .tag = .assignment,
                    .x = .{ .assignment = .{
                        .name = t.ident,
                        .value = value,
                    } },
                };
                const semi = it.next().?;
                if (semi != .semi) {
                    try semi.print();
                    @panic("\nparse: Unknown Lexeren following expression statement ^");
                }
                return assign_stmt;
            } else {
                const s = Statement{
                    .tag = .expression,
                    .x = .{ .expression = try parseExpression(ally, it, t) },
                };
                const semi = it.next().?;
                if (semi != .semi) {
                    try semi.print();
                    @panic("\nparse: Unknown Lexeren following expression statement ^");
                }
                return s;
            }
        },
        .inline_batch => |batch| {
            return .{
                .tag = .inline_batch,
                .x = .{ .inline_batch = batch },
            };
        },
        .openCurly => {
            var statements = try std.ArrayList(Statement).initCapacity(ally, 4);
            var stmt = try parseStatement(ally, it);
            while (stmt.tag != .eof) {
                try printStatement(stmt);
                try statements.append(stmt);
                stmt = try parseStatement(ally, it);
            }
            const closecurly = Lexer.peek(it).?;
            if (closecurly != .closeCurly) {
                try std.io.getStdOut().writer().print("{}\n", .{closecurly});
                @panic("\nparse: Unknown Lexeren following block ^");
            }
            _ = it.next().?; // }
            const block = try ally.create(Block);
            block.* = .{ .statements = try statements.toOwnedSlice() };
            return .{
                .tag = .block,
                .x = .{ .block = block },
            };
        },
        .openParen,
        .closeParen,
        .closeCurly,
        .semi,
        .comma,
        .colon,
        .equal,
        .excl,
        .star,
        .plus,
        .hyphen,
        .slash,
        .percent,
        .unknown,
        => {
            it.* = snapshot; // restore
            return .{ .tag = .eof };
        },
    }
    unreachable;
}

pub fn parse(ally: std.mem.Allocator, it: *Lexer) !Program {
    var statements = try std.ArrayList(Statement).initCapacity(ally, 16);
    var stmt = try parseStatement(ally, it);
    while (stmt.tag != .eof) {
        try statements.append(stmt);
        stmt = try parseStatement(ally, it);
    }
    return .{ .statements = try statements.toOwnedSlice() };
}
